File: labs-opensource/M08-conversation-management/solution/smart_manager.py
"""
M08 Lab - Step 3: SmartConversationManager — SOLUTION
======================================================
Run: python smart_manager.py
"""

import json
import time
from pathlib import Path
from typing import Optional

from openai import OpenAI

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")


class SmartConversationManager:
    """Full-featured manager with summarization and persistence."""

    def __init__(
        self,
        system_prompt: str = "You are a helpful assistant.",
        model: str = "mistral",
        token_threshold: int = 8_000,
        recent_turns_to_keep: int = 4,
    ):
        self.system_prompt = system_prompt
        self.model = model
        self.messages: list = []
        self.token_threshold = token_threshold
        self.recent_turns_to_keep = recent_turns_to_keep
        self.summary: Optional[str] = None
        self.summary_history: list = []
        self.last_input_tokens = 0
        self.total_input_tokens = 0
        self.total_output_tokens = 0

    def _should_summarize(self) -> bool:
        return self.last_input_tokens > self.token_threshold

    def _summarize_old_messages(self) -> None:
        """Use Mistral to summarize older messages."""
        keep_count = self.recent_turns_to_keep * 2  # user+assistant pairs
        if len(self.messages) <= keep_count:
            return

        old_messages = self.messages[:-keep_count]
        recent_messages = self.messages[-keep_count:]

        summary_prompt = (
            "Summarize this conversation concisely. "
            "Preserve: key decisions, user preferences, important facts. "
            "Skip: greetings, filler.\n\n"
        )
        for msg in old_messages:
            summary_prompt += f"{msg['role']}: {msg['content']}\n"

        try:
            response = client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "You are a conversation summarizer. Be concise."},
                    {"role": "user", "content": summary_prompt},
                ],
            )
            new_summary = response.choices[0].message.content

            if self.summary:
                new_summary = f"Previous context: {self.summary}\n\nRecent: {new_summary}"

            self.summary = new_summary
            self.summary_history.append({
                "timestamp": time.time(),
                "messages_summarized": len(old_messages),
            })

            self.messages = [
                {"role": "user", "content": f"[Conversation summary: {self.summary}]"},
                {"role": "assistant", "content": "Understood. I have the conversation context."},
                *recent_messages,
            ]
        except Exception:
            # Graceful fallback: plain truncation. Never crash the conversation.
            self.messages = recent_messages

    def send(self, user_message: str) -> str:
        """Send with automatic summarization when needed."""
        self.messages.append({"role": "user", "content": user_message})

        try:
            response = client.chat.completions.create(
                model=self.model,
                messages=[{"role": "system", "content": self.system_prompt}] + self.messages,
            )
            if response.usage:
                self.last_input_tokens = response.usage.prompt_tokens
                self.total_input_tokens += response.usage.prompt_tokens
                self.total_output_tokens += response.usage.completion_tokens

            assistant_text = response.choices[0].message.content
            self.messages.append({"role": "assistant", "content": assistant_text})

            if self._should_summarize():
                self._summarize_old_messages()

            return assistant_text

        except Exception as e:
            self.messages.pop()
            raise RuntimeError(f"API call failed: {e}") from e

    # ── Persistence ──
    def save(self, filepath: str) -> None:
        state = {
            "system_prompt": self.system_prompt,
            "model": self.model,
            "messages": self.messages,
            "summary": self.summary,
            "summary_history": self.summary_history,
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens,
            "saved_at": time.time(),
        }
        Path(filepath).write_text(json.dumps(state, indent=2), encoding="utf-8")

    @classmethod
    def load(cls, filepath: str) -> "SmartConversationManager":
        data = json.loads(Path(filepath).read_text(encoding="utf-8"))
        mgr = cls(system_prompt=data["system_prompt"], model=data["model"])
        mgr.messages = data["messages"]
        mgr.summary = data.get("summary")
        mgr.summary_history = data.get("summary_history", [])
        mgr.total_input_tokens = data.get("total_input_tokens", 0)
        mgr.total_output_tokens = data.get("total_output_tokens", 0)
        return mgr


if __name__ == "__main__":
    manager = SmartConversationManager(
        token_threshold=3_000,
        recent_turns_to_keep=4,
        system_prompt="You are a helpful coding assistant.",
    )

    turns = [
        "Help me build a REST API with FastAPI. Explain the project structure in detail.",
        "Now explain how to add a database with SQLAlchemy, including models and migrations.",
        "Add JWT authentication with refresh tokens. Show the full flow.",
        "How do I write tests for all of this with pytest?",
        "What about deployment — Docker, environment variables, the works?",
    ]
    for i, t in enumerate(turns, 1):
        manager.send(t)
        print(f"Turn {i}: last_input={manager.last_input_tokens} tokens, "
              f"stored={len(manager.messages)} msgs")

    print(f"\nSummaries created: {len(manager.summary_history)}")

    manager.save("conversation_state.json")
    restored = SmartConversationManager.load("conversation_state.json")
    print(f"Restored: {len(restored.messages)} messages, summary={'yes' if restored.summary else 'no'}")
    print(f"\nRestored agent says: {restored.send('Where were we? One sentence.')[:200]}")

File: labs-opensource/M08-conversation-management/solution/smart_manager.js
/**
 * M08 Lab - Step 3: SmartConversationManager — SOLUTION
 * ======================================================
 * Run: node smart_manager.js
 */

import OpenAI from "openai";
import { readFileSync, writeFileSync } from "node:fs";

import { pathToFileURL } from "node:url";
const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });

export class SmartConversationManager {
  constructor({
    systemPrompt = "You are a helpful assistant.",
    model = "mistral",
    tokenThreshold = 8_000,
    recentTurnsToKeep = 4,
  } = {}) {
    this.systemPrompt = systemPrompt;
    this.model = model;
    this.messages = [];
    this.tokenThreshold = tokenThreshold;
    this.recentTurnsToKeep = recentTurnsToKeep;
    this.summary = null;
    this.summaryHistory = [];
    this.lastInputTokens = 0;
    this.totalInputTokens = 0;
    this.totalOutputTokens = 0;
  }

  _shouldSummarize() {
    return this.lastInputTokens > this.tokenThreshold;
  }

  async _summarizeOldMessages() {
    const keepCount = this.recentTurnsToKeep * 2; // user+assistant pairs
    if (this.messages.length <= keepCount) return;

    const oldMessages = this.messages.slice(0, -keepCount);
    const recentMessages = this.messages.slice(-keepCount);

    let summaryPrompt =
      "Summarize this conversation concisely. " +
      "Preserve: key decisions, user preferences, important facts. " +
      "Skip: greetings, filler.\n\n";
    for (const msg of oldMessages) {
      summaryPrompt += `${msg.role}: ${msg.content}\n`;
    }

    try {
      const response = await client.chat.completions.create({
        model: this.model,
        messages: [
          { role: "system", content: "You are a conversation summarizer. Be concise." },
          { role: "user", content: summaryPrompt },
        ],
      });
      let newSummary = response.choices[0].message.content;

      if (this.summary) {
        newSummary = `Previous context: ${this.summary}\n\nRecent: ${newSummary}`;
      }

      this.summary = newSummary;
      this.summaryHistory.push({
        timestamp: Date.now(),
        messagesSummarized: oldMessages.length,
      });

      this.messages = [
        { role: "user", content: `[Conversation summary: ${this.summary}]` },
        { role: "assistant", content: "Understood. I have the conversation context." },
        ...recentMessages,
      ];
    } catch {
      // Graceful fallback: plain truncation. Never crash the conversation.
      this.messages = recentMessages;
    }
  }

  async send(userMessage) {
    this.messages.push({ role: "user", content: userMessage });

    try {
      const response = await client.chat.completions.create({
        model: this.model,
        messages: [{ role: "system", content: this.systemPrompt }, ...this.messages],
      });

      if (response.usage) {
        this.lastInputTokens = response.usage.prompt_tokens;
        this.totalInputTokens += response.usage.prompt_tokens;
        this.totalOutputTokens += response.usage.completion_tokens;
      }

      const assistantText = response.choices[0].message.content;
      this.messages.push({ role: "assistant", content: assistantText });

      if (this._shouldSummarize()) {
        await this._summarizeOldMessages();
      }

      return assistantText;
    } catch (error) {
      this.messages.pop();
      throw new Error(`API call failed: ${error.message}`);
    }
  }

  // ── Persistence ──
  save(filepath) {
    const state = {
      systemPrompt: this.systemPrompt,
      model: this.model,
      messages: this.messages,
      summary: this.summary,
      summaryHistory: this.summaryHistory,
      totalInputTokens: this.totalInputTokens,
      totalOutputTokens: this.totalOutputTokens,
      savedAt: Date.now(),
    };
    writeFileSync(filepath, JSON.stringify(state, null, 2));
  }

  static load(filepath) {
    const data = JSON.parse(readFileSync(filepath, "utf-8"));
    const mgr = new SmartConversationManager({
      systemPrompt: data.systemPrompt,
      model: data.model,
    });
    mgr.messages = data.messages;
    mgr.summary = data.summary ?? null;
    mgr.summaryHistory = data.summaryHistory ?? [];
    mgr.totalInputTokens = data.totalInputTokens ?? 0;
    mgr.totalOutputTokens = data.totalOutputTokens ?? 0;
    return mgr;
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const manager = new SmartConversationManager({
    tokenThreshold: 3_000,
    recentTurnsToKeep: 4,
    systemPrompt: "You are a helpful coding assistant.",
  });

  const turns = [
    "Help me build a REST API with FastAPI. Explain the project structure in detail.",
    "Now explain how to add a database with SQLAlchemy, including models and migrations.",
    "Add JWT authentication with refresh tokens. Show the full flow.",
    "How do I write tests for all of this with pytest?",
    "What about deployment — Docker, environment variables, the works?",
  ];
  for (let i = 0; i < turns.length; i++) {
    await manager.send(turns[i]);
    console.log(`Turn ${i + 1}: last_input=${manager.lastInputTokens} tokens, stored=${manager.messages.length} msgs`);
  }

  console.log(`\nSummaries created: ${manager.summaryHistory.length}`);

  manager.save("conversation_state.json");
  const restored = SmartConversationManager.load("conversation_state.json");
  console.log(`Restored: ${restored.messages.length} messages, summary=${restored.summary ? "yes" : "no"}`);
  console.log(`\nRestored agent says: ${(await restored.send("Where were we? One sentence.")).slice(0, 200)}`);
}
